### PR TITLE
opt to validate incoming data before proxy is executed

### PR DESCRIPTION
**Pull request by kovalenp**
Provide an option to preValidate incoming data (check out: https://github.com/fastify/fastify-http-proxy/issues/311)

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by evahteev — APPROVED** [review]
LGTM

File: index.js [hunk 1]
```diff
@@ -260,11 +260,15 @@ async function fastifyHttpProxy (fastify, opts) {
...
-  if (opts.proxyPayloads !== false) {
+  if (opts.preValidation === undefined && opts.proxyPayloads !== false) {
     fastify.addContentTypeParser('application/json', bodyParser)
     fastify.addContentTypeParser('*', bodyParser)
   }
 
+  if (opts.preValidation) {
+    fastify.addHook('preValidation', opts.preValidation)
+  }
+
```
Comment (Uzlopak) [inline index.js:271]: Why do we not proxy Payloads if preValidation is set?

Comment (kovalenp) [reply index.js:271]: `req.body` is not parsed and comes as a stream if:

```
fastify.addContentTypeParser('application/json', bodyParser)
fastify.addContentTypeParser('*', bodyParser)
```

condition is executed, as described here: https://github.com/fastify/fastify-http-proxy/issues/86#issue-591840299

actually, the payload is passed in this case, see:

https://github.com/fastify/fastify-http-proxy/pull/312/files#diff-a561630bb56b82342bc66697aee2ad96efddcbc9d150665abd6fb7ecb7c0ab2fR269

Comment (Uzlopak) [inline index.js:271]:
```suggestion
  if (opts.preValidation) {
    fastify.addHook('preValidation', opts.preValidation)
  } else if (opts.proxyPayloads !== false) {
    fastify.addContentTypeParser('application/json', bodyParser)
    fastify.addContentTypeParser('*', bodyParser)
  }

```

**Review by mcollina — CHANGES_REQUESTED** [review]
I'm -1 to add this option.

This can be achieved with the following:

```js
app.register(async function (app) {
  app.addHook('preValidation', ...)
  app.register(httpProxy, {
    proxyPayloads: false
  })
})
```

Could you add that to the documentation instead?

**Comment by kovalenp** [comment]
thanks @mcollina did you check https://github.com/fastify/fastify-http-proxy/issues/311 ? It is on the TODO list.

what if I don't want to run preValidation on every route just on the proxied one?

**Comment by mcollina** [comment]
> thanks @mcollina did you check https://github.com/fastify/fastify-http-proxy/issues/311 ? 

I did now, I still think what you want to achieve can be done with the approach I described in https://github.com/fastify/fastify-http-proxy/pull/312#pullrequestreview-1540665638. Note that I set `proxyPayloads` to `false, which does the trick and it enables body validation.

> It is on the TODO list.

I assembled that a long time ago, I might have been mistaken in adding this.

> what if I don't want to run preValidation on every route just on the proxied one?

The snippet in https://github.com/fastify/fastify-http-proxy/pull/312#pullrequestreview-1540665638 adds it only to the proxied routes.

**Comment by kovalenp** [comment]
I still find adding `preValidator` option a more convenient way (and it's consistent to preHandler option), I'd rather remove the `proxyPayloads` instead. 

But it's indeed up to the team. Feel free to close the PR if you decide that it's not the preferable way.

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by mcollina** [comment]
Can you fix the failing tests?

**Comment by kovalenp** [comment]
> Can you fix the failing tests?

I wish I could! It seems that tests are executed on different OS images, and two failing ones are related to test coverage:

```
ERROR: Coverage for lines (91.24%) does not meet global threshold (100%)
ERROR: Coverage for functions (93.75%) does not meet global threshold (100%)
ERROR: Coverage for branches (90.24%) does not meet global threshold (100%)
ERROR: Coverage for statements (90.84%) does not meet global threshold (100%)
```

how it's setup? how it's possible the same set of tests shows different coverage on different agents?

**Comment by mcollina** [comment]
For whatever reason this is actually breaking websocket support as this test PR is passing: https://github.com/fastify/fastify-http-proxy/pull/315

**Comment by kovalenp** [comment]
> For whatever reason this is actually breaking websocket support as this test PR is passing: #315

This PR didn't trigger unit tests though, there are only two checks. I'm positive main branch have the issue

**Comment by mcollina** [comment]
Verified, it's a problem.

**Comment by mcollina** [comment]
please rebase, I've fixed it

**Comment by kovalenp** [comment]
> please rebase, I've fixed it

Nice! Respect Sir. Rebase is done

File: types/index.d.ts [hunk 2]
```diff
@@ -31,6 +31,7 @@ declare namespace fastifyHttpProxy {
     proxyPayloads?: boolean;
     preHandler?: preHandlerHookHandler;
     beforeHandler?: preHandlerHookHandler;
+    preValidation?: preValidationHookHandler;
```
Comment (mcollina) [inline types/index.d.ts:34]: Can you add a test for this type in https://github.com/fastify/fastify-http-proxy/blob/master/types/index.test-d.ts?

Comment (kovalenp) [reply types/index.d.ts:34]: Addressed here: 0c0e505c7180d3f31d51e2c081d334ba7da580d5

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by cb1kenobi** [comment]
Excellent work! I can't wait to use it. Mind pushing a new release? Thanks!

**Comment by denchen** [comment]
Hi. Can we get a release for this? I saw `preValidation` in the docs and assumed that I could use it, but the latest release does not have this feature.

**Comment by Eomm** [comment]
Released as 9.3.0 👍🏼

**Pull request merged**